onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -7,63 +7,82 @@
 #include "core/framework/op_kernel.h"
 #include "core/framework/ml_value.h"
 
+using namespace ::onnxruntime::common;
+
 namespace onnxruntime {
 
-Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified, bool& deleted) {
-  // Create execution frame for executing constant nodes.
-  OptimizerExecutionFrame::Info info({&node}, graph.GetAllInitializedTensors());
+Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-  std::vector<int> fetch_mlvalue_idxs;
-  for (const auto* node_out : node.OutputDefs()) {
-    fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-  }
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
+      continue;
+    }
 
-  OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level));
 
-  auto* kernel = info.GetKernel(node.Index());
-  OpKernelContext op_kernel_context(&frame, kernel, ::onnxruntime::logging::LoggingManager::DefaultLogger());
+    // Check if constant folding can be applied on this node.
+    if (!graph_utils::IsSupportedProvider(*node, GetCompatibleExecutionProviders()) ||
+        excluded_op_types_.find(node->OpType()) != excluded_op_types_.end() ||
+        // constant folding is not currently supported for control flow operators.
+        graph_utils::ContainsSubgraph(*node) ||
+        // if the node output is in the graph output, we will get a graph with no nodes.
+        // TODO check if this is allowed in ONNX and ORT.
+        graph.IsNodeOutputsInGraphOutputs(*node) ||
+        !graph_utils::AllNodeInputsAreConstant(graph, *node)) {
+      continue;
+    }
 
-  kernel->Compute(&op_kernel_context);
+    // Create execution frame for executing constant nodes.
+    OptimizerExecutionFrame::Info info({node}, graph.GetAllInitializedTensors());
 
```

Comment:
If we setup a merged list of initializers for each graph (get list of current graph and add initializers from parent graphs if they don't exist in the local list) at the start of ApplyImpl, and add the implicit inputs from the parent node in OptimizerExecutionFrame::Info::Info,  I think it should be possible to handle subgraphs more completely. 

Did you want to attempt that here, or add it in a separate PR?

Reply:
I will do it as a separate one to create proper tests for constant folding in subgraphs too.